rust/src/parser.rs

```diff
@@ -0,0 +1,308 @@
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+// written by Pierre Chifflier  <chifflier@wzdftpd.net>
+
+//! Parser registration functions and common interface
+
+use core::{DetectEngineState,Flow,AppLayerEventType,AppLayerDecoderEvents,STREAM_TOCLIENT,STREAM_TOSERVER};
+
+use log::*;
+
+use libc;
+use libc::{c_void,c_char,c_int};
+use std::ffi::CString;
+
+
+// Application layer protocol identifiers (app-layer-protos.h)
+
+pub const ALPROTO_UNKNOWN : AppProto = 0;
+pub const ALPROTO_FAILED : AppProto = 20;
+
+/// Rust parser declaration
+pub struct RustParser {
+    /// Parser name.
+    pub name:            &'static str,
+    /// Default port
+    pub default_port:    &'static str,
+
+    /// IP Protocol (libc::IPPROTO_UDP, libc::IPPROTO_TCP, etc.)
+    pub ipproto:         c_int,
+
+    /// Protocol name.
+    pub proto_name:      &'static str,
+
+    /// Probing function, for packets going to server
+    pub probe_ts:        ProbeFn,
+    /// Probing function, for packets going to client
+    pub probe_tc:        ProbeFn,
+
+    /// Minimum frame depth for probing
+    pub min_depth:       u16,
+    /// Maximum frame depth for probing
+    pub max_depth:       u16,
+
+    /// Allocation function for a new state
+    pub state_new:       StateAllocFn,
+    /// Function called to free a state
+    pub state_free:      StateFreeFn,
+
+    /// Parsing function, for packets going to server
+    pub parse_ts:        ParseFn,
+    /// Parsing function, for packets going to client
+    pub parse_tc:        ParseFn,
+
+    /// Get the current transaction count
+    pub get_tx_count:    StateGetTxCntFn,
+    /// Get a transaction
+    pub get_tx:          StateGetTxFn,
+    /// Function called to free a transaction
+    pub tx_free:         StateTxFreeFn,
+    /// Function returning the current transaction completion status
+    pub tx_get_comp_st:  StateGetTxCompletionStatusFn,
+    /// Function returning the current transaction progress
+    pub tx_get_progress: StateGetProgressFn,
+
+    /// Logged transaction getter function
+    pub get_tx_logged:   Option<GetTxLoggedFn>,
+    /// Logged transaction setter function
+    pub set_tx_logged:   Option<SetTxLoggedFn>,
+
+    /// Function called to get a detection state
+    pub get_de_state:    GetDetectStateFn,
+    /// Function called to set a detection state
+    pub set_de_state:    SetDetectStateFn,
+    /// Function to check if a detection state is present
+    pub has_de_state:    Option<HasDetectStateFn>,
+
+    /// Function to check if there are events
+    pub has_events:      Option<HasEventsFn>,
+    /// Function to get events
+    pub get_events:      Option<GetEventsFn>,
+    /// Function to get an event description
+    pub get_eventinfo:   Option<GetEventInfoFn>,
+
+    // Missing:
+    // - LocalStorageAlloc / LocalStorageFree
+    // - SetTxMpmIDs / GetTxMpmIDs
+    // - GetFiles
```

Comment:
the files are a bit trick between rust and C right now. Do you see any potential issues?

Reply:
Not sure I understand what you mean.
C and Rust files are quite similar and must be kept in sync (and moving the registration to C adds the declaration of the structure in the C code as well). Hopefully, the structures will not change often.
If we need to do the declaration automatically, it's possible to use rust-bindgen.

For the functions not wrapped, it's not a problem: we can add the declaration of the C functions (for ex `AppLayerParserRegisterGetFilesFunc`) and call the after registering the parser. Or we can add a field to the structure.

Reply:
Maybe it's not a problem at all, just wanted to mention that it's a bit of a hack now :)